go/vt/wrangler/materializer_test.go

```diff
@@ -2831,55 +2835,251 @@ func TestMaterializerManyToManySomeUnreachable(t *testing.T) {
+				},
+			},
+			getStreamInsert: getStreamInsert,
 		},
 		{
-			targetShards: []string{"0"},
+			name:         "unsharded source, sharded target, use all source shards",
+			targetShards: []string{"-80", "80-"},
+			sourceShards: []string{"-"},
+			insertMap: map[string][]string{
+				"-80": {"-"},
+				"80-": {"-"},
+			},
+			sourceVSchema: &vschemapb.Keyspace{
+				Sharded: false,
+			},
+			targetVSchema:   targetVSchema,
+			getStreamInsert: getStreamInsert,
+		},
+		{
+			name:         "sharded source, unsharded target, use all source shards",
+			targetShards: []string{"-"},
 			sourceShards: []string{"-80", "80-"},
-			insertMap:    map[string][]string{"0": {"-80", "80-"}},
+			insertMap: map[string][]string{
+				"-": {"-80", "80-"},
+			},
+			targetVSchema: &vschemapb.Keyspace{
+				Sharded: false,
+			},
+			getStreamInsert: func(sourceShard, targetShard string) string {
+				return fmt.Sprintf(`.*shard:\\"%s\\" filter:{rules:{match:\\"t1\\" filter:\\"select.*t1`, sourceShard)
+			},
 		},
 		{
+			name:         "target secondary vindexes, use intersecting source shards",
 			targetShards: []string{"-80", "80-"},
-			sourceShards: []string{"0"},
-			insertMap:    map[string][]string{"-80": {"0"}, "80-": {"0"}},
+			sourceShards: []string{"-40", "40-80", "80-c0", "c0-"},
+			insertMap:    map[string][]string{"-80": {"-40", "40-80"}, "80-": {"80-c0", "c0-"}},
+			targetVSchema: &vschemapb.Keyspace{
+				Sharded: true,
+				Vindexes: map[string]*vschemapb.Vindex{
+					"hash": {
+						Type: "hash",
+					},
+					"lookup_vdx": {
+						Type: "lookup",
+					},
+				},
+				Tables: map[string]*vschemapb.Table{
+					"t1": {
+						ColumnVindexes: []*vschemapb.ColumnVindex{
+							{
+								Column: "c1",
+								Name:   "hash",
+							},
+							{
+								Column: "c2",
+								Name:   "lookup_vdx",
+							},
+						},
+					},
+				},
+			},
+			getStreamInsert: getStreamInsert,
+		},
+		{
+			name:         "same vindex type but different cols, use all source shards",
+			targetShards: []string{"-80", "80-"},
+			sourceShards: []string{"-40", "40-80", "80-c0", "c0-"},
+			sourceVSchema: &vschemapb.Keyspace{
+				Sharded: true,
+				Vindexes: map[string]*vschemapb.Vindex{
+					"hash": {
+						Type: "hash",
+					},
+				},
+				Tables: map[string]*vschemapb.Table{
+					"t1": {
+						ColumnVindexes: []*vschemapb.ColumnVindex{{
+							Column: "c1",
+							Name:   "hash",
+						}},
+					},
+				},
+			},
+			targetVSchema: &vschemapb.Keyspace{
+				Sharded: true,
+				Vindexes: map[string]*vschemapb.Vindex{
+					"hash": {
+						Type: "hash",
+					},
+				},
+				Tables: map[string]*vschemapb.Table{
+					"t1": {
+						ColumnVindexes: []*vschemapb.ColumnVindex{{
+							Column: "c2",
+							Name:   "hash",
+						}},
+					},
+				},
+			},
+			getStreamInsert: func(sourceShard, targetShard string) string {
+				return fmt.Sprintf(`.*shard:\\"%s\\" filter:{rules:{match:\\"t1\\" filter:\\"select.*t1`, sourceShard)
+			},
```

Comment:
OK, here it's w/o any keyrange filtering. Might be worth a comment just to make that easier to see.

Reply:
I don't really know what to make of the fact that there isn't any keyrange filtering here. It makes sense to me that this workflow produces a set of many-to-many source-to-target streams, but I would still expect there to be keyrange filtering so that each target shard ends up with a distinct subset of source data. Does this seem right to you, or did I possibly introduce a bug 🤔

Reply:
> I don't really know what to make of the fact that there isn't any keyrange filtering here. It makes sense to me that this workflow produces a set of many-to-many source-to-target streams, but I would still expect there to be keyrange filtering so that each target shard ends up with a distinct subset of source data. Does this seem right to you, or did I possibly introduce a bug 🤔

I just run this test on your PR by logging the rules that the `in_keyrange` is present. The reason it works without you specifying it in the match string above is that we append it with the `insertPrefix` which  is prefixed with `/` and hence results in a 'starts with' match. 

But we should probably add the 'in_keyrange' to the match strings above, like in the other tests.

Reply:
Ah thank you, will do that